Added print human readable snapshot to base instrument class

AdriaanRol opened the pull request:
Fixes #391 

Changes proposed in this pull request:
- Add a print_readable_snapshot to the base instrument class 


@giulioungaretti 
@nulinspiratie, I think you said you liked this. 

Below two pictures to compare the conventional snapshot and the readable version. 
Note that the readable snapshot is not intended to replace the existing snapshot function. 

## Readable snapshot:
![screenshot 2016-12-20 13 21 20](https://cloud.githubusercontent.com/assets/6142932/21350380/52a2388a-c6b7-11e6-8605-b1b0cdd835ee.png)

## Conventional snapshot: 
![screenshot 2016-12-20 13 21 31](https://cloud.githubusercontent.com/assets/6142932/21350378/4f70d3b0-c6b7-11e6-9d5b-1afa70b7834f.png)

jenshnielsen commented on qcodes/instrument/base.py:
```diff
@@ -434,6 +434,15 @@ def snapshot_base(self, update=False):
                 snap[attr] = getattr(self, attr)
         return snap
 
+    def print_readable_snapshot(self, update=False):
```
Should we add a docstring explaining that this returns name, value and units for all parameters

AdriaanRol replied:
@jenshnielsen I think that would be a good idea. I probably won't have time to do that today. Feel free to add it if you want to make it a quick merge, otherwise I'll probably get to it somewhere in the holidays.

jenshnielsen replied:
Other than than I think it looks ready to merge

MerlinSmiles commented:
would it make sense to return that formatted string instead of printing?

also maybe have a dynamic width of the columns depending on content, i did something like that in the __repr__ of the dataset... https://github.com/QCoDeS/Qcodes/blob/master/qcodes/data/data_set.py#L865

jenshnielsen commented:
@MerlinSmiles No objections from me but that would require a different method name.

MerlinSmiles commented:
@jenshnielsen sure 😄

nulinspiratie commented:
Very useful functionality, it definitely makes it easier to look at the state of an instrument.

@MerlinSmiles What about a kwarg return_string, which is set to False by default?

AdriaanRol commented:
@MerlinSmiles , I really like the idea of the dynamic width. 

I don't like the idea of the function returning a formatted string. The reason is that this is a convenience function that should be readily accessible, without the overhead of having to go through `print(instr.readable_snapshot())`. I have nothing against the idea of @nulinspiratie of adding a kwarg `return_string` to optionally allow returning the formatted string. 

Would that suit you?

giulioungaretti commented:
I would recommend against a kwarg ! 
This PR and issue #391  are part of the bigger picture issue: the metadata are mostly useless for human consumption (especially while actually measuring).

MerlinSmiles commented:
I'm not a big fan of a kwarg in that case either, but also I have in general no strong opinions on this.
Maybe have a `print_readable_snapshot()` and a `readable_snapshot()` ^^ that would make `print(readable_snapshot()) almost the same amount of typing :D
But im not sure if that name makes so much sense as there is information left out. maybe have it be a `short` or `quick_snapshot()`

AdriaanRol commented:
Ok, based on the discussion I will 
- [ ] Add a docstring
- [ ] Add a dynamic width 

I will not add return formatted string option.

giulioungaretti commented:
sound like a great plan @AdriaanRol ! 🎅

nulinspiratie commented:
@AdriaanRol Yeah sounds good, would it also make sense to add the units directly after a value instead of a separate column, but only if units is not None?

nulinspiratie commented:
@AdriaanRol would it be hard to extend this, such that you could for instance also do station.print_readable_snapshot(), which would then print all its instruments and parameters? That could provide you a readable full list of all the info of instruments & parameters. I personally wouldn't mind having that saved as a separate file during a measurement.

Also, what do you think of print_snapshot instead of print_readable_snapshot?

AdriaanRol commented:
@nulinspiratie , I like the extension you propose. However, for the moment I'll stick to the proposed simple changes as I find myself quite busy, though maybe I'll do it as it looks very simple on the surface, no commitment though sorry :( . 

As to your idea of saving this info, that is exactly what save metadata should do, and also what the snapshot is intended for. I know that we use it for that purpose in any case.

AdriaanRol commented:
I've updated the print_readable_snapshot function. 

The output now looks like below. 
It starts by printing the instrument name, followed by the table headers and a newline. 
Every parameter is printed at a fixed width and right aligned. 
If a value is a floating point type it will use a :.5g general formatter to make the values more readable. A proper unit prefix would be preferable but that is beyond the scope of this PR. 
units will only be added if they exist. 
```
Simulation parameters:
           parameter 	value
--------------------------------------------------------------------------------
                  E_c:	3e+08 (Hz)
                  IDN:	None
                  J_2:	7.0711e+07 (Hz)
            asymmetry:	0 (a.u.)
 dac_flux_coefficient:	1 (a.u.)
             f_01_max:	5.94e+09 (Hz)
        f_interaction:	4.8e+09 (Hz)
        lambda_coeffs:	[1 0 0] (a.u.)
         pulse_length:	2e-08 (s)
        sampling_rate:	2e+09 (Hz)
              theta_f:	1.5708 (rad)

```

AdriaanRol commented:
Edit: @giulioungaretti , I'd say this PR is ready for review and otherwise done from my part.

nataliejpg commented:
This looks really useful @AdriaanRol, quick question (just for personal preference), is the right_align hard coded? Also as an extension (not to get in the way of this pull request being merged or but just to put the thought out there) it would be fab if you could optionally specify a subset of parameters from the station to avoid printing the whole thing. I often find myself wanting to check the same 6 or 7 parameters and there isn't currently a nice way to do this. It would be a sort of precursor to having a 'monitor'.  Perhaps in the next generation of this function...  Thanks for this though 👍

AdriaanRol commented:
@nataliejpg ,
> is the right_align hard coded?

Yes it is. I figured it looked best to have everything aligned but I can see that twice left aligned would also work. Do you think that would look better? 
I don't think adding many options to this function would be a good idea, that kind of defeats the purpose of having a quick print statement. 

 > optionally specify a subset of parameters from the station to avoid printing the whole thing.

I think that would be great, however at the moment I do not see an elegant way to do this. I'm thinking along the lines of changing the representation of the snapshot (which basically acts like a dict anyway) so that you can select the keys you want. If I'd have this I would not want to explicitly type the names I want but rather provide some string parsing magic. In any case cool ideas but a bit out of scope at this moment :) 

>  It would be a sort of precursor to having a 'monitor'. Perhaps in the next generation of this function... 

Some kind of monitor (I'm thinking using pyqtgraph or other pyqt widgets) that periodically question the snapshot is becoming more and more important I'd say. Quickly hacking something together is quite possible based on the snapshots but I'm much looking forward to whatever the core devs come up with after they finish the dataset.

AdriaanRol commented:
I've updated the print_readable snapshot a bit to implement some of the improvements that where discussed in this thread. 
- @MerlinSmiles I've implemented variable width for the parameter name field 
- @nataliejpg I've decided that left aligning (as you suggested) looks better 
- I've made the value line cut of after (by default) 80 characters so that an array like parameter that you won't read easily like this anyway, won't break the usability of this function. 

@giulioungaretti @jenshnielsen @WilliamHPNielsen , could one of you review this so it can be merged? 


```
Simulation parameters:
	parameter           value
--------------------------------------------------------------------------------
E_c                  :	3e+08 (Hz)
IDN                  :	None
J_2                  :	7.0711e+07 (Hz)
asymmetry            :	0 (a.u.)
dac_flux_coefficient :	1 (a.u.)
f_01_max             :	5.94e+09 (Hz)
f_interaction        :	4.8e+09 (Hz)
lambda_coeffs        :	[ 0.28318615  0.63683849  0.06542197  0.41768607  0.35...
pulse_length         :	2e-08 (s)
sampling_rate        :	2e+09 (Hz)
theta_f              :	1.5708 (rad)
```

giulioungaretti commented:
LGTM! Merge on monday

nataliejpg commented:
@AdriaanRol looks great!
